alibi_detect/od/_mahalanobis.py

```diff
@@ -0,0 +1,178 @@
+    ) -> None:
+        """
+        The Mahalanobis outlier detection method.
+
+        The Mahalanobis computes the directions of variation of a dataset and uses them to detect when points are
+        outliers by checking to see if the point vary from dataset points in unexpected ways.
+
+        When we fit the Mahalanobis method we compute the covariance matrix of the reference data and its eigenvectors
+        and eigenvalues. We filter small eigenvalues for numerical stability using the `min_eigenvalue` parameter. We
+        then inversely weight each eigenvector by its eigenvalue.
+
+        When we score test points we project them onto the eigenvectors and compute the l2-norm of the projected point.
+        Because the eigenvectors are inversely weighted by the eigenvalues, the score will take into account the
+        difference in variance along each direction of variation. If a test point lies along a direction of high
+        variation then it must lie very far out to obtain a high score. If a test point lies along a direction of low
+        variation then it doesn't need to lie very far out to obtain a high score.
+
+        Parameters
+        ----------
+        min_eigenvalue
+            Eigenvectors with eigenvalues below this value will be discarded. This is to ensure numerical stability.
+        backend
+            Backend used for outlier detection. Defaults to ``'pytorch'``. Options are ``'pytorch'``.
+        device
+            Device type used. The default tries to use the GPU and falls back on CPU if needed. Can be specified by
+            passing either ``'cuda'``, ``'gpu'``, ``'cpu'`` or an instance of ``torch.device``.
+
+        Raises
+        ------
+        NotImplementedError
+            If choice of `backend` is not implemented.
+        """
+        super().__init__()
+
+        backend_str: str = backend.lower()
+        BackendValidator(
+            backend_options={'pytorch': ['pytorch']},
+            construct_name=self.__class__.__name__
+        ).verify_backend(backend_str)
+
+        backend_cls = backends[backend]
+        self.backend = backend_cls(min_eigenvalue, device=device)
+
+        # set metadata
+        self.meta['detector_type'] = 'outlier'
+        self.meta['data_type'] = 'numeric'
+        self.meta['online'] = False
+
+    def fit(self, x_ref: np.ndarray) -> None:
+        """Fit the detector on reference data.
+
+        Fitting the Mahalanobis method amounts to computing the covariance matrix and its eigenvectors. We filter out
+        very small eigenvalues using the `min_eigenvalue` parameter. We then scale the eigenvectors such that the data
+        projected onto them has mean ``0`` and std ``1``.
+
+        Parameters
+        ----------
+        x_ref
+            Reference data used to fit the detector.
+        """
+        self.backend.fit(self.backend._to_tensor(x_ref))
+
+    @catch_error('NotFittedError')
```

Comment:
If the detector has not had a threshold inferred the score function can still be used because we don't require a threshold to compute the scores. If the predict method is used but the threshold has not been inferred we return a dictionary containing the scores and leave the outlier predictions blank.

This behaviour is the same as the kNN case if you look closely. If the kNN is not an ensemble then it behaves the same as the above. If its an ensemble then the `score` method requires the ensemble to be fit which is done in the `infer_threshold` method, not the `fit` method. Also, see [this](https://github.com/SeldonIO/alibi-detect/pull/677#issuecomment-1439881214) comment.

Reply:
Btw is this the same behaviour as our current detectors and is it documented? It's hard to process predictions downstream if they are allowed to lack predictions if `infer_threshold` has not been set. @arnaudvl

Reply:
ok, noted. It's easy to revert if we want to. Basically, we just throw an error [here](https://github.com/SeldonIO/alibi-detect/blob/b7f2228d39320ef51e5ff42cd558249e7a4d998b/alibi_detect/od/pytorch/base.py#L163).  Currently the old outlier detectors will presumably throw an error [here](https://github.com/SeldonIO/alibi-detect/blob/b7f2228d39320ef51e5ff42cd558249e7a4d998b/alibi_detect/od/mahalanobis.py#L344) as `self.threshold=None` although I haven't actually tested this.